[Lim Jan Jay] iP

jaysmyname opened the pull request:
# ConnorBot

> "Free your mind." - Zenyatta, *Overwatch*

Tired of always having to remember your deadlines and duties? :thinking: Well fret no more, because **ConnorBot** is here!
**ConnorBot** features state-of-the-art technology that makes your life that much simpler! 

It has functionality for (but not limited to):
- taking in different types of tasks, like deadlines and events,
- marking and unmarking your tasks,
- recognizing date formats,
- searching for your tasks with keywords

All you need to do to set it up is:
1. Download **ConnorBot** [here](https://github.com/jaysmyname/ip/releases)
2. Copy the jar file into an empty folder
3. Open a command window in that folder
4. Run the command java `-jar ip_project.jar`
5. ...and you're good to go! Go ahead and add your tasks and start improving the quality of your life.

Future features that are planned be implemented:
- [ ] GUI for ConnorBot (Coming Soon)
- [ ] Sort tasks by date (Coming Soon)
- [ ] Reminders (Coming Soon)


If you're a Java programmer, you can use it to practice Java yourself! The `main` method is here:
```java
public static void main(String[] args) {
    try {
        new Connor(TASK_FILEPATH).run();
    } catch (IOException e) {
        print(e.getMessage());
    }
}
```

jaysmyname committed:
Renamed Duke.java to Connor.java, added Greet, Echo & Exit

jaysmyname committed:
Added text storing, changed access levels of methods and variables from public to private

jaysmyname committed:
Added Task.java and the ability to mark/unmark tasks

jaysmyname committed:
Added classes ToDo, Event and Deadline

jaysmyname committed:
Changed valueOf methods to parseInt

jaysmyname committed:
Added functionality for ToDos, Deadlines and Events

jaysmyname committed:
Minor code cleanup

jaysmyname committed:
Added addTask method to simplify switch cases

jaysmyname committed:
Renamed task list from xs to taskList

jaysmyname committed:
General code cleanup

jaysmyname committed:
Added Override

jaysmyname committed:
Added semi-automated text UI testing

jaysmyname committed:
Added errors when empty descriptions are given for tasks

jaysmyname committed:
Added errors when mark/unmark index is not an integer or out of range

jaysmyname committed:
Added contractions to bot's speech

jaysmyname committed:
Interpret method trims before splitting

jaysmyname committed:
Added ability to delete tasks, and a few methods

jaysmyname committed:
Added enum TaskType.java

jaysmyname committed:
Made most Connor messages constants for easier access

jaysmyname committed:
Add ConnorTest

jaysmyname committed:
Improve interpret method so statement array won't throw ArrayIndexOutOfBoundsException

jaysmyname committed:
Add enum functionality for addTask method

jaysmyname committed:
Add saving functionality, new Exceptions, and clear command

jaysmyname committed:
Add date recognition for Deadlines and Events

jaysmyname committed:
Merge branch 'branch-Level-7'

jaysmyname committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Connor.java
#	src/main/java/Deadline.java
#	src/main/java/Event.java

Resolution: keep both branch implementations

jaysmyname committed:
Make the code more OOP

jaysmyname committed:
Rename MarkStatus to TaskStatus

jaysmyname committed:
Divide classes into packages

jaysmyname committed:
Remove ConnorTest.java

jaysmyname committed:
Add equals method override

jaysmyname committed:
Add JUnit Tests for Storage and TaskList

jaysmyname committed:
Change constant LINE

jaysmyname committed:
Package the App as a JAR file

jaysmyname committed:
Change version number to reflect on JAR release

jaysmyname committed:
Update EXPECTED.txt file

jaysmyname committed:
Cleanup code

jaysmyname committed:
Improve flexibility of filePath directory

jaysmyname committed:
Add Javadoc to several classes and their methods

jaysmyname committed:
Add Javadoc to several classes and their methods

jaysmyname committed:
Tweak code to comply with given coding standard

jaysmyname committed:
Add Find feature to allow users to find tasks with a keyword

jaysmyname committed:
Merge branch 'branch-A-Javadoc'

jaysmyname committed:
Merge branch 'branch-A-CodingStandard'

jaysmyname committed:
Merge branch 'branch-Level-9'

jaysmyname committed:
Fix bug in Parser.java

jaysmyname committed:
Update automate text UI testing

jaysmyname committed:
Improve findTasks method to be case-insensitive

jaysmyname committed:
Add Gradle support to project

jaysmyname committed:
Set up gradle to run unit tests

jaysmyname committed:
Merge branch 'branch-A-Gradle'

jaysmyname committed:
Add Checkstyle

jaysmyname committed:
Make code follow new checkstyle

jaysmyname committed:
Merge branch 'branch-A-CheckStyle'

KeithCZW commented on src/main/java/connor/Connor.java:
```diff
@@ -0,0 +1,113 @@
+ * @version 1.0.0
+ */
+public class Connor {
+    public static final String CURRENT_VERSION = "Version 1.0.0";
+    public static final String LINE = "_".repeat(66);
+    public static final String INDENT = " ".repeat(4);
+    public static final String INPUT_HERE = ">>> ";
```
I like that all your static variables are all in full caps, which is the trend among all your files. Good job!

KeithCZW commented on src/main/java/connor/Connor.java:
```diff
@@ -0,0 +1,113 @@
+
+    private static final String ERROR_FILE_NOT_FOUND = "Error! Task file not found!";
+
+    private static boolean isActive = true;
+    private static final Scanner sc = new Scanner(System.in);
+
+    private Storage storage;
+    private Ui ui;
```
Variables are appropriately named. All using camel case and booleans have the is prefix.

KeithCZW commented on src/main/java/connor/Ui.java:
```diff
@@ -0,0 +1,41 @@
+     * Prints the Connor logo and user greeting.
+     */
+    public void greetings() {
+        String logo = " .d8888b.\n"
+                + "d88P  Y88b\n"
+                + "888    888\n"
+                + "888         .d88b.  88888b.  88888b.   .d88b.  888d888\n"
+                + "888        d88\"\"88b 888 \"88b 888 \"88b d88\"\"88b 888P\"\n"
+                + "888    888 888  888 888  888 888  888 888  888 888\n"
+                + "Y88b  d88P Y88..88P 888  888 888  888 Y88..88P 88\n"
+                + " \"Y8888P\"   \"Y88P\"  888  888 888  888  \"Y88P\"  888\n";
+        print("\n" + logo + "\n" + Connor.CURRENT_VERSION);
+        print(Connor.LINE);
+        print("Hi, my name is Connor! I'm your personalised android assistant.\n"
+                + "Loading your current tasks...");
+        print(Connor.LINE);
```
Nice graphic touch!

KeithCZW commented on src/main/java/connor/command/AddCommand.java:
```diff
@@ -0,0 +1,27 @@
+ * Represents an Add {@code Command}.
+ */
+public class AddCommand extends Command {
+    private TaskType tt;
```
Perhaps a bit too general naming for tt?

jaysmyname replied:
Perhaps, I'll most likely change it in the next push

KeithCZW commented on src/main/java/connor/task/TaskList.java:
```diff
@@ -0,0 +1,269 @@
+ * Stores and edits the task list.
+ */
+public class TaskList {
+    private static final String INDENT = Connor.INDENT;
+    private static final String LINE = Connor.LINE;
+    private static final String EMPTY_TASK_LIST = "Your task list is empty.";
+    private static final String SHOW_TASKS = "Here are your current tasks: ";
+    private static final String ADD_NEW_TASK = "Alright, I've added a new task: ";
+    private static final String DELETE_TASK = "Alright, I've deleted the task: ";
+    private static final String CLEAR_TASKS_CONFIRM = "Are you REALLY sure you want to clear all your tasks?\n"
+            + "Type 'yes' to confirm, type anything else (or nothing) to cancel.";
+    private static final String CLEAR_TASKS_CONFIRMED = "Poof! All your tasks are cleared!";
+    private static final String CLEAR_TASKS_CANCEL = "Cancelled clearing all tasks. Phew!";
+    private static final String MARK_TASK = "Good job! I've marked the following task as completed: ";
+    private static final String UNMARK_TASK = "Understood. I've unmarked the following task: ";
+    private static final String NO_MATCHING_TASKS = "Sorry, there are no matching tasks in your list.";
+    private static final String SHOW_MATCHING_TASKS = "Here are the matching tasks in your list: ";
+    private static final String ERROR_EMPTY_INDEX = "Error! Index cannot be empty.";
+    private static final String ERROR_EMPTY_TASK_DESC = "Error! Tasks cannot have an empty description.";
+    private static final String ERROR_EMPTY_DL_DESC = "Error! Deadlines cannot have empty descriptions or dates.";
+    private static final String ERROR_EMPTY_EVENT_DESC = "Error! Events cannot have empty descriptions or dates.";
+    private static final String ERROR_INDEX_NOT_INTEGER = "Error! Index must be a valid integer.";
+    private static final String ERROR_INDEX_OUT_OF_RANGE = "Error! Given index is out of range.";
+    private static final String ERROR_INVALID_DL_FORMAT = "Error! Wrong format for deadlines.\n\n"
+            + "Deadline tasks must include \"by\" in the description.\nExample: \n"
+            + ">>> deadline Finish project /by Monday Morning";
+    private static final String ERROR_INVALID_EVENT_FORMAT = "Error! Wrong format for events.\n\n"
+            + "Event tasks must include \"at\" in the description.\nExample: \n"
+            + ">>> event Birthday Party at May 5th";
+    private static final String ERROR_INVALID_TASK_TYPE = "Oh no! Incorrect Task type!";
+    private static final String ERROR_MARK_EMPTY = "Error! I can't mark an empty task list!";
+    private static final String ERROR_UNMARK_EMPTY = "Error! I can't unmark an empty task list!";
```
I like that you encapsulate every error message into one file, to make it easier to edit them in the future.

jaysmyname replied:
Thank you! I agree, it is very convenient, especially if the message appears multiple times.

KeithCZW reviewed (commented):
Overall, your PR is looking very good! I struggled very hard to find any issues with the coding standard.

jaysmyname committed:
Add Gui in progress

jaysmyname committed:
Add functionality for GUI

jaysmyname committed:
Increase maxHeight and maxWidth of some nodes

jaysmyname committed:
Fix Label not wrapping text

jaysmyname committed:
Make GUI look nicer

jaysmyname committed:
Add JavaDoc for relevant GUI classes and methods.

jaysmyname committed:
Merge branch 'branch-Level-10'

ianfromdover commented on text-ui-test/EXPECTED.TXT:
```diff
@@ -1,7 +1,161 @@
+    3. [E][X] Birthday Party (At: Saturday Afternoon)
+    4. [E][ ] Meet Friend (At: Sunday 12pm)
+__________________________________________________________________
+>>> clear
+__________________________________________________________________
+Poof! All your tasks are cleared!
+__________________________________________________________________
```
Just a usability thing, it would be good to ask the user for confirmation before this potentially irreversible task is carried out to prevent the user from losing progress :)

ianfromdover commented on src/main/java/connor/task/ToDo.java:
```diff
@@ -0,0 +1,47 @@
+package connor.task;
+
+/**
+ * Represents a {@code Task} of type {@code ToDo}.
+ */
+public class ToDo extends Task {
```
Thanks for teaching me how to use the `@code` tag in my JavaDocs 👍

ianfromdover commented on src/main/java/connor/task/TaskList.java:
```diff
@@ -0,0 +1,312 @@
+    private static final String UNMARK_TASK = "Understood. I've unmarked the following task: ";
+    private static final String NO_MATCHING_TASKS = "Sorry, there are no matching tasks in your list.";
+    private static final String SHOW_MATCHING_TASKS = "Here are the matching tasks in your list: ";
+    private static final String ERROR_EMPTY_INDEX = "Error! Index cannot be empty.";
+    private static final String ERROR_EMPTY_TASK_DESC = "Error! Tasks cannot have an empty description.";
+    private static final String ERROR_EMPTY_DL_DESC = "Error! Deadlines cannot have empty descriptions or dates.";
+    private static final String ERROR_EMPTY_EVENT_DESC = "Error! Events cannot have empty descriptions or dates.";
+    private static final String ERROR_INDEX_NOT_INTEGER = "Error! Index must be a valid integer.";
+    private static final String ERROR_INDEX_OUT_OF_RANGE = "Error! Given index is out of range.";
+    private static final String ERROR_INVALID_DL_FORMAT = "Error! Wrong format for deadlines.\n\n"
+            + "Deadline tasks must include \"by\" in the description.\nExample: \n"
+            + ">>> deadline Finish project /by Monday Morning";
+    private static final String ERROR_INVALID_DL_FORMAT_GUI = "Error! Wrong format for deadlines.\n\n"
+            + "Deadline tasks must include \"by\" in the description.\nExample: \n"
+            + "'deadline Finish project /by Monday Morning'";
+    private static final String ERROR_INVALID_EVENT_FORMAT = "Error! Wrong format for events.\n\n"
+            + "Event tasks must include \"at\" in the description.\nExample: \n"
+            + ">>> event Birthday Party at May 5th";
+    private static final String ERROR_INVALID_EVENT_FORMAT_GUI = "Error! Wrong format for events.\n\n"
+            + "Event tasks must include \"at\" in the description.\nExample: \n"
+            + "'event Birthday Party at May 5th'";
+    private static final String ERROR_INVALID_TASK_TYPE = "Oh no! Incorrect Task type!";
+    private static final String ERROR_INVALID_TASK_STATUS = "Oh no! Invalid Task status!";
+    private static final String ERROR_MARK_EMPTY = "Error! I can't mark an empty task list!";
+    private static final String ERROR_UNMARK_EMPTY = "Error! I can't unmark an empty task list!";
```
Perhaps the error constants can be declared in the `ConnorException` class instead :)

ianfromdover reviewed (commented):
LGTM. One of the most well-organised projects i've seen, easy to read and understand, and your teammates are very lucky to have you with them 👍

jaysmyname committed:
Add assertions to TaskList and Storage

jaysmyname committed:
Storage class taskToString & stringToTask methods handles tasks
case-by-case on its own.

The methods are too long and unwieldy which can cause difficulty in
finding bugs later on.

Extracting out the methods based on the task type allows us to clearly
see what is being done based on the task's type.

This is to make the code look cleaner and make bug-finding easier.

jaysmyname committed:
Edit methods to fit camelCase format

jaysmyname committed:
Extract methods from markStatus

jaysmyname committed:
Follow Test CheckStyle

jaysmyname committed:
Extract methods in Parser

jaysmyname committed:
Merge pull request #1 from jaysmyname/branch-A-CodeQuality

Code Quality Improvement

jaysmyname committed:
Merge pull request #2 from jaysmyname/branch-A-Assertions

Add assertions to TaskList and Storage

jaysmyname committed:
Merge branch 'branch-A-CodeQuality'

jaysmyname committed:
Set up CI

jaysmyname committed:
Make Deadline and Event tasks take certain date and time format only

jaysmyname committed:
Add sorting command and functionality

Program does not have a way to sort tasks.

Adding a way to sort tasks helps tidy up Deadline and Event tasks.

There are two ways to sort tasks:
* by chronological order
* by task type

jaysmyname committed:
Merge branch 'branch-C-Sort'

jaysmyname committed:
Minor bug fixes

jaysmyname committed:
Minor bug fixes

jaysmyname committed:
Improve the GUI visually

jaysmyname committed:
Update version

jaysmyname committed:
Give title to window

jaysmyname committed:
Add Ui example

jaysmyname committed:
Update README

jaysmyname committed:
Update README

jaysmyname committed:
Update README

jaysmyname committed:
Set theme jekyll-theme-midnight

jaysmyname committed:
Update User Guide

jaysmyname committed:
Merge branch 'master' of https://github.com/jaysmyname/ip

jaysmyname committed:
Set theme jekyll-theme-cayman

jaysmyname committed:
Update README in project root folder

jaysmyname committed:
Update User Guide

jaysmyname committed:
Fix bug in TaskList

jaysmyname committed:
Fix error in UG
